File: Linked List /Python/4. 19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int count = 0;
        ListNode* curr = head;
        while(curr != NULL)
        {
            count += 1;
            curr = curr->next;
        }
        int k = count - n;  // Denoting the node to be deleted
        curr = head;
        if(k == 0)
        {
            ListNode* temp = head;
            head = head-> next;
            delete temp;
            return head;
        }

        for(int i = 0; i < k - 1; i++)
        {
            curr = curr->next;        // Traversing till the node before that node to be deleted 
        } 
        ListNode *temp = curr->next;
        curr->next = curr->next->next;
        delete temp;
        return head;
    }
};

File: Linked List /Python/4. 19. Remove Nth Node from end of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        temp = head
        count = 0
        while temp is not None:
            count += 1
            temp = temp.next

 # Denoting the index of the node from the start to be removed at the given position from the end
        k = count - n    
        curr = head
        if (k == 0):
            temp = head
            head = head.next
            del temp     # remove the first node
            return head
        for i in range(0, k-1):
            curr = curr.next
        
        temp = curr.next
        curr.next = curr.next.next
        del temp        # delete the given node  
        return head
